aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -337,7 +336,6 @@ SparseTensor& copy_sparse_(SparseTensor& self, const SparseTensor& src, bool non
 
 SparseTensor coalesce_sparse_cpu(const SparseTensor& self) {
   AT_ASSERT(self.defined());
-  AT_ASSERT(!self.is_variable());  // TODO: change this to check `.requires_grad()` and `GradMode::is_enabled()` when Variable and Tensor are merged
```

Comment:
If we look at the original PR where this check was added, the only purpose of it was to make sure that `_get_sparse_impl(self)` at https://github.com/pytorch/pytorch/pull/8409/files#diff-8b62faad691b7ba86dde89b61d0cae1fR336 is being run on a Tensor instead of a Variable (since running it on a Variable will return Variable::Impl instead of TensorImpl, which will make the subsequent operation incorrect).

However, the current version of this function doesn't have the `_get_sparse_impl(self)` call anymore, which is why we no longer need this check. One other place where this check can matter is `SparseTensor dst = self.clone()` below, as having the check can prevent `self.clone()` from building up autograd history. I suggested that we change it to `SparseTensor dst = self.detach().clone()`, to prevent building autograd history regardless of whether `self` requires grad or not.